File: implementation/scipamato/core/web/src/test/java/ch/difty/scipamato/core/web/newsletter/NewsletterTopicDefinitionProviderTest.java
package ch.difty.scipamato.core.web.newsletter;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import org.apache.wicket.extensions.markup.html.repeater.data.sort.SortOrder;
import org.apache.wicket.model.IModel;
import org.apache.wicket.util.tester.WicketTester;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatcher;
import org.mockito.Mock;

import ch.difty.scipamato.common.persistence.paging.PaginationContext;
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicDefinition;
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicFilter;
import ch.difty.scipamato.core.web.AbstractWicketTest;

class NewsletterTopicDefinitionProviderTest extends AbstractWicketTest {

    private NewsletterTopicDefinitionProvider provider;

    @Mock
    private NewsletterTopicFilter     filterMock;
    @Mock
    private NewsletterTopicDefinition entityMock;

    private List<NewsletterTopicDefinition> papers;

    @BeforeEach
    void setUp() {
        new WicketTester(application);
        provider = new NewsletterTopicDefinitionProvider(filterMock);

        papers = Arrays.asList(entityMock, entityMock, entityMock);
    }

    @AfterEach
    void tearDown() {
        verifyNoMoreInteractions(newsletterTopicServiceMock, entityMock);
    }

    @Test
    void defaultFilterIsNewNewsletterTopicFilter() {
        provider = new NewsletterTopicDefinitionProvider();
        assertThat(provider.getFilterState()).isEqualToComparingFieldByField(new NewsletterTopicFilter());
    }

    @Test
    void nullFilterResultsInNewNewsletterTopicFilter() {
        NewsletterTopicDefinitionProvider p = new NewsletterTopicDefinitionProvider(null);
        assertThat(p.getFilterState()).isEqualToComparingFieldByField(new NewsletterTopicFilter());
    }

    @Test
    void size() {
        int size = 5;
        when(newsletterTopicServiceMock.countByFilter(filterMock)).thenReturn(size);
        assertThat(provider.size()).isEqualTo(size);
        verify(newsletterTopicServiceMock).countByFilter(filterMock);
    }

    @Test
    void gettingModel_wrapsEntity() {
        IModel<NewsletterTopicDefinition> model = provider.model(entityMock);
        assertThat(model.getObject()).isEqualTo(entityMock);
    }

    @Test
    void gettingFilterState_returnsFilter() {
        assertThat(provider.getFilterState()).isEqualTo(filterMock);
    }

    @Test
    void settingFilterState() {
        provider = new NewsletterTopicDefinitionProvider();
        assertThat(provider.getFilterState()).isNotEqualTo(filterMock);
        provider.setFilterState(filterMock);
        assertThat(provider.getFilterState()).isEqualTo(filterMock);
    }

    private class PaginationContextMatcher implements ArgumentMatcher<PaginationContext> {

        private final int    pageSize;
        private final String sort;

        @SuppressWarnings("SameParameterValue")
        PaginationContextMatcher(int pageSize, String sort) {
            this.pageSize = pageSize;
            this.sort = sort;
        }

        @Override
        public boolean matches(PaginationContext p) {
            return p != null && p.getPageSize() == pageSize && sort.equals(p
                .getSort()
                .toString());
        }
    }

    @Test
    void iterating_withNoRecords_returnsNoRecords() {
        papers = Collections.emptyList();
        when(newsletterTopicServiceMock.findPageOfEntityDefinitions(eq(filterMock),
            isA(PaginationContext.class))).thenReturn(papers.iterator());
        Iterator<NewsletterTopicDefinition> it = provider.iterator(0, 3);
        assertThat(it.hasNext()).isFalse();
        verify(newsletterTopicServiceMock).findPageOfEntityDefinitions(eq(filterMock),
            argThat(new NewsletterTopicDefinitionProviderTest.PaginationContextMatcher(3, "title: ASC")));
    }

    @Test
    void iterating_throughFirst() {
        when(newsletterTopicServiceMock.findPageOfEntityDefinitions(eq(filterMock),
            isA(PaginationContext.class))).thenReturn(papers.iterator());
        Iterator<NewsletterTopicDefinition> it = provider.iterator(0, 3);
        assertRecordsIn(it);
        verify(newsletterTopicServiceMock).findPageOfEntityDefinitions(eq(filterMock),
            argThat(new NewsletterTopicDefinitionProviderTest.PaginationContextMatcher(3, "title: ASC")));
    }

    private void assertRecordsIn(Iterator<NewsletterTopicDefinition> it) {
        int i = 0;
        while (i++ < 3) {
            assertThat(it.hasNext()).isTrue();
            it.next();
        }
        assertThat(it.hasNext()).isFalse();
    }

    @Test
    void iterating_throughSecondPage() {
        when(newsletterTopicServiceMock.findPageOfEntityDefinitions(eq(filterMock),
            isA(PaginationContext.class))).thenReturn(papers.iterator());
        Iterator<NewsletterTopicDefinition> it = provider.iterator(3, 3);
        assertRecordsIn(it);
        verify(newsletterTopicServiceMock).findPageOfEntityDefinitions(eq(filterMock),
            argThat(new NewsletterTopicDefinitionProviderTest.PaginationContextMatcher(3, "title: ASC")));
    }

    @Test
    void iterating_throughThirdPage() {
        provider.setSort("title", SortOrder.DESCENDING);
        when(newsletterTopicServiceMock.findPageOfEntityDefinitions(eq(filterMock),
            isA(PaginationContext.class))).thenReturn(papers.iterator());
        Iterator<NewsletterTopicDefinition> it = provider.iterator(6, 3);
        assertRecordsIn(it);
        verify(newsletterTopicServiceMock).findPageOfEntityDefinitions(eq(filterMock),
            argThat(new NewsletterTopicDefinitionProviderTest.PaginationContextMatcher(3, "title: DESC")));
    }

}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/newsletter/NewsletterTopicDefinitionProviderTest.kt
package ch.difty.scipamato.core.web.newsletter

import ch.difty.scipamato.common.persistence.paging.matchPaginationContext
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicDefinition
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicFilter
import ch.difty.scipamato.core.web.AbstractWicketTest
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.verify
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldNotBeEqualTo
import org.apache.wicket.extensions.markup.html.repeater.data.sort.SortOrder
import org.apache.wicket.util.tester.WicketTester
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class NewsletterTopicDefinitionProviderTest : AbstractWicketTest() {

    private lateinit var provider: NewsletterTopicDefinitionProvider

    private val filterDummy = NewsletterTopicFilter().apply { titleMask = "foo" }
    private val entityDummy = NewsletterTopicDefinition(null, "en", null)

    private val entities = listOf(entityDummy, entityDummy, entityDummy)

    @BeforeEach
    fun setUp() {
        WicketTester(application)
        provider = NewsletterTopicDefinitionProvider(filterDummy)
    }

    @AfterEach
    fun tearDown() {
        confirmVerified(newsletterTopicServiceMock)
    }

    @Test
    fun defaultFilterIsNewNewsletterTopicFilter() {
        provider = NewsletterTopicDefinitionProvider()
        provider.filterState shouldBeEqualTo NewsletterTopicFilter()
    }

    @Test
    fun nullFilterResultsInNewNewsletterTopicFilter() {
        val p = NewsletterTopicDefinitionProvider(null)
        p.filterState shouldBeEqualTo NewsletterTopicFilter()
    }

    @Test
    fun size() {
        val size = 5
        every { newsletterTopicServiceMock.countByFilter(filterDummy) } returns size
        provider.size() shouldBeEqualTo size.toLong()
        verify { newsletterTopicServiceMock.countByFilter(filterDummy) }
    }

    @Test
    fun gettingModel_wrapsEntity() {
        val model = provider.model(entityDummy)
        model.getObject() shouldBeEqualTo entityDummy
    }

    @Test
    fun gettingFilterState_returnsFilter() {
        provider.filterState shouldBeEqualTo filterDummy
    }

    @Test
    fun settingFilterState() {
        provider = NewsletterTopicDefinitionProvider()
        provider.filterState shouldNotBeEqualTo filterDummy
        provider.filterState = filterDummy
        provider.filterState shouldBeEqualTo filterDummy
    }

    @Test
    fun iterating_withNoRecords_returnsNoRecords() {
        every { newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, any()) } returns emptyList<NewsletterTopicDefinition>().iterator()
        val it = provider.iterator(0, 3)
        it.hasNext().shouldBeFalse()
        verify {
            newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, matchPaginationContext(0, 3, "title: ASC"))
        }
    }

    @Test
    fun iterating_throughFirst() {
        every { newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, any()) } returns entities.iterator()
        val it = provider.iterator(0, 3)
        assertRecordsIn(it)
        verify {
            newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, matchPaginationContext(0, 3, "title: ASC"))
        }
    }

    private fun assertRecordsIn(it: Iterator<NewsletterTopicDefinition>) {
        repeat(3) { _ ->
            it.hasNext().shouldBeTrue()
            it.next()
        }
        it.hasNext().shouldBeFalse()
    }

    @Test
    fun iterating_throughSecondPage() {
        every { newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, any()) } returns entities.iterator()
        val it = provider.iterator(3, 3)
        assertRecordsIn(it)
        verify {
            newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, matchPaginationContext(3, 3, "title: ASC"))
        }
    }

    @Test
    fun iterating_throughThirdPage() {
        provider.setSort("title", SortOrder.DESCENDING)
        every { newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, any()) } returns entities.iterator()
        val it = provider.iterator(6, 3)
        assertRecordsIn(it)
        verify {
            newsletterTopicServiceMock.findPageOfEntityDefinitions(filterDummy, matchPaginationContext(6, 3, "title: DESC"))
        }
    }
}
